plugins/elanfp/fu-elanfp-device.c

```diff
@@ -403,6 +404,19 @@ fu_elanfp_device_write_firmware(FuDevice *device,
 		return FALSE;
 	if (!fu_elanfp_device_write_payload(self, payload, fu_progress_get_child(progress), error))
 		return FALSE;
+
+	/* hardware reset */
+	if (!fu_elanfp_device_do_xfer(self,
+				      (guint8 *)&usb_buf,
+				      sizeof(usb_buf),
+				      NULL,
+				      0,
+				      TRUE,
+				      NULL,
+				      error)) {
+		g_prefix_error(error, "failed to hardware reset ");
+		return FALSE;
+	}
 	fu_progress_step_done(progress);
```

Comment:
don't you also need to set `fu_device_add_flag(device, FWUPD_DEVICE_FLAG_WAIT_FOR_REPLUG);` here and `fu_device_add_private_flag(FU_DEVICE(self), FU_DEVICE_PRIVATE_FLAG_ONLY_WAIT_FOR_REPLUG);` in `fu_elanfp_device_init`?

Reply:
ohh, just in case it's helpful, if you `git push --force` this branch it'll update this PR automatically without having to open a new one.

Reply:
> don't you also need to set `fu_device_add_flag(device, FWUPD_DEVICE_FLAG_WAIT_FOR_REPLUG);` here and `fu_device_add_private_flag(FU_DEVICE(self), FU_DEVICE_PRIVATE_FLAG_ONLY_WAIT_FOR_REPLUG);` in `fu_elanfp_device_init`?

elanfp is internal device, end-user can't replug it. So we will use command for the hardware reset.

Reply:
Ahh, FWUPD_DEVICE_FLAG_WAIT_FOR_REPLUG doesn't ask the user to replug, it just tells the daemon that a USB disconnect and reconnect is expected t happen.

Reply:
OK. We will add it and update again.

Reply:
Add "fu_device_add_private_flag(FU_DEVICE(self), FU_DEVICE_PRIVATE_FLAG_ONLY_WAIT_FOR_REPLUG);" in fu_elanfp_device_init only.

Will get failure if add FWUPD_DEVICE_FLAG_WAIT_FOR_REPLUG. The error message as following:
error message: failed to prepare composite action: failed to wait for composite prepare: device 102a4d416cd1b9e16372a5569555299bd41eef1d did not come back

Reply:
Did you put FWUPD_DEVICE_FLAG_WAIT_FOR_REPLUG at the end of fu_elanfp_device_write_firmware? The FU_DEVICE_PRIVATE_FLAG_ONLY_WAIT_FOR_REPLUG is the only flag that you set in init.